Bb chapter 4 (#695)
mlr-org/mlr3book: book/chapters/chapter4/hyperparameter_optimization.qmd

### Tuning Instance with `ti` {#sec-tuning-instance}

The tuning instance collects the tuner-agnostic information required to optimize a model, i.e., all information about the tuning process, except for the tuning algorithm itself.
This includes the task to tune over, the learner to tune, the resampling method and measure used to analytically compare hyperparameter optimization configurations, and the terminator to determine when the measure has been optimized 'enough'.

> This implicitly defines a "black box" objective function, mapping hyperparameter configurations to (stochastic) performance values, that is to be optimized.

Reviewer: ```suggestion
This implicitly defines a "black box" objective function, mapping hyperparameter configurations to (stochastic) performance values, to be optimized.
```

This concept will be revisited in @sec-optimization-advanced.

A `r index("tuning instance")` can be constructed explicitly with the `r ref("ti()")` function, or we can tune a learner with the `r ref("tune()")` function, which implicitly creates a tuning instance, as shown in @sec-autotuner.
We cover the `ti()` approach first as this allows finer control of tuning and a more nuanced discussion about the design and use of `mlr3tuning`.

Continuing our example, we will construct a `r index("single-objective")` tuning problem (i.e., tuning over *one* measure) by using the `ti()` function to create a `r ref("TuningInstanceSingleCrit")`, we will return to `r index('multi-objective tuning')` in @sec-multi-metrics-tuning.

For this example, we will use three-fold CV and optimize the classification error measure.

> Note that in the next section, we will continue our example with a grid search tuner, so we select `trm("none")`, as we want to iterate over the full grid without preliminary stopping.

Reviewer: ```suggestion
Note that in the next section, we will continue our example with a grid search tuner, so we select `trm("none")` below as we will want to iterate over the full grid without stopping too soon.
```

Hyperband is discussed in full detail in @sec-hyperband.

Other implemented algorithms for numeric search spaces are Generalized Simulated Annealing [@xiang2013; @tsallis1996] and various nonlinear optimization algorithms.

#### Choosing strategies {.unnumbered .unlisted}

As a rule of thumb, if the search space is small or does not have a complex structure, grid search may be able to exhaustively evaluate the entire search space in a reasonable time. However, `r index('grid search')` is generally not recommended due to the curse of dimensionality -- the grid size 'blows up' very quickly as the number of parameters to tune increases -- and insufficient coverage of numeric search spaces.
By construction, grid search cannot evaluate a large number of unique values per hyperparameter, which is suboptimal when some hyperparameters have minimal impact on performance while others do.
In such scenarios, `r index('random search')` is often a better choice as it considers more unique values per hyperparameter compared to grid search.

> For higher-dimensional search spaces or search spaces with more complex structure, more guided optimization algorithms such as evolutionary strategies or Bayesian optimization tend to perform better and are more likely to result in peak performance. When choosing between `r index('evolutionary strategies')` and `r index('Bayesian optimization', lower = FALSE)`, the cost of function evaluation is highly relevant. If hyperparameter configurations can be evaluated quickly, evolutionary strategies often work well. On the other hand, if model evaluations are time-consuming and the optimization budget is limited, Bayesian optimization is usually preferred, as it is quite sample efficient compared to other algorithms, i.e., much less function evaluations are needed to find good configurations.

Reviewer: ```suggestion
For higher-dimensional search spaces or search spaces with more complex structure, more guided optimization algorithms such as evolutionary strategies or Bayesian optimization tend to perform better and are more likely to result in peak performance.
When choosing between `r index('evolutionary strategies')` and `r index('Bayesian optimization', lower = FALSE)`, the cost of function evaluation is highly relevant.
If hyperparameter configurations can be evaluated quickly, evolutionary strategies often work well.
On the other hand, if model evaluations are time-consuming and the optimization budget is limited, Bayesian optimization is usually preferred, as it is quite sample efficient compared to other algorithms, i.e., less function evaluations are needed to find good configurations.
Hence, Bayesian optimization is usually recommended for HPO.
```

> Due to this latter point, Bayesian optimization is usually recommended for HPO.

Reviewer: ```suggestion
```

You can access all the resamplings combined in a `r ref("BenchmarkResult")` object with `instance$archive$benchmark_result`.

> Finally, to visualize the results, you can use `r ref("mlr3viz::autoplot.TuningInstanceSingleCrit()", "autoplot()")` (@fig-surface).

Reviewer: ```suggestion
Finally, to visualize the results, you can use `r ref("mlr3viz::autoplot.TuningInstanceSingleCrit")` (@fig-surface).
```

lrn_svm_tuned$train(tsk_sonar)$model
```

> ## Convenience with `tune` and `auto_tuner` {#sec-autotuner}

Reviewer: ```suggestion
## Convenient Tuning with `tune` and `auto_tuner` {#sec-autotuner}
```

If you have the resources we recommend utilizing parallelization when tuning (@sec-parallelization).
:::

A common mistake is to think of nested resampling as a method to select optimal model configurations.
Nested resampling is a method to compare models and to estimate the generalization performance of a tuned model, however, this is the performance based on multiple different configurations (one from each outer fold) and not performance based on a *single* configuration (@sec-resample-overfitting).
If you are interested in identifying optimal configurations, then use `r ref("tune()", index = TRUE)`/`r ref("ti()")` or `r ref("auto_tuner()", index = TRUE)` with `$train()` on the complete dataset.

> ### Nested Resampling by Resampling an `AutoTuner`

Reviewer: ```suggestion
### Nested Resampling with an `AutoTuner`
```

  .(iteration, cost, gamma, classif.ce)]
```

### The Right (and Wrong) Way to Estimate Performance {#sec-resample-overfitting}

{{< include ../../common/_optional.qmd >}}

> In this short section we will empirically demonstrate that directly reporting tuning performance without nested resampling results in optimistically biased performance estimates. In this experiment we tune several parameters from `lrn("classif.xgboost"). To best estimate the generalization performance we make use of the `"moons"` `r ref("TaskGenerator", aside = TRUE)`.

Reviewer: ```suggestion
In this short section we will empirically demonstrate that directly reporting tuning performance without nested resampling results in optimistically biased performance estimates.
In this experiment we tune several parameters from `lrn("classif.xgboost")`.
To best estimate the generalization performance we make use of the `"moons"` `r ref("TaskGenerator", aside = TRUE)`.
```
